File: 498-diagonal-traverse/diagonal-traverse.py
# Optimal Code
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        # number of rows and column
        m = len(mat)
        n = len(mat[0])
        # dictionary to map : (row + col) -> element on the diagonal
        mp = {}
        # store the final diagonal traversal
        result = []
        # group all matrix eleemnt by their diagonal index (row + col)
        for i in range(m):
            for j in range(n):
                # cells having the same values of (row + col) belong to the same diagonal
                key = i + j
                # create a new list for the diagonal if its does not exits
                if key not in mp:
                    mp[key] = []
                # store the current element
                mp[key].append(mat[i][j])
        # even diagonal are traversed upward odd diagonal downward
        flip = True
        # process diagonals in increasing order
        for key in sorted(mp.keys()):
            # reverse every alternate diagonal to obtain the required zigzg order
            if flip:
                mp[key].reverse()
            # add the current diagonal to the final answer
            result.extend(mp[key])
            # toggle the traversal direaction for the next diagonal
            flip = not flip
        # return the diagonal traversal
        return result 

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 498-diagonal-traverse/diagonal-traverse.cpp
// Brute Force Code
class Solution {
public:
    // return the diagonal traversal of the matrix
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        // number of rows and columns
        int rows = mat.size();
        int cols = mat[0].size();
        // store the final diagonal traversal
        vector<int> result;
        // total number of diagonals in the matrix
        int total_diagonals = rows + cols - 1;
        // process each diagonal
        for (int diagonal = 0; diagonal < total_diagonals; diagonal++) {
            // store current diagonal elements
            vector<int> current_diagonal;
            // collect all elements belonging to this diagonal
            for (int row = 0; row < rows; row++) {
                for (int col = 0; col < cols; col++) {
                    // cells with the same value of (row + col) belong to the same diagonal
                    if (row + col == diagonal) {
                        current_diagonal.push_back(mat[row][col]);
                    }
                }
            }
            // even-numbered diagonals are traversed from bottom to top, so reverse their order
            if (diagonal % 2 == 0) {
                reverse(current_diagonal.begin(),current_diagonal.end());
            }
            // append the current diagonal to the result
            result.insert(result.end(), current_diagonal.begin(),current_diagonal.end());
        }
        // return the diagonal traversal
        return result;
    }
};

// Time Complexity : O(N^3)
// Space Complexity : O(N)